Friendly fire fixes and sam sites

pevers opened the pull request:
This PR will fix the following issues:

https://github.com/OpenRA/OpenRA/issues/7442

^ is solved by clearing the turret activities when not being able to attack (during low power).

https://github.com/OpenRA/OpenRA/issues/7403

^ is being solved by clearing the targets when the owner is changed.

pevers committed:
fixed taking over sam/pillbox/etc. to stop firing

fixed issue with powered down samsites

more optimization

very small optimization

undo changes to powerdown samsite fixes

refixed powered down sam sites

removed debug line

redid the fix in another approach by queueing a sell activity on the top of the queue. This was already present but didn't always work

fixed line I removed

prevented attack activity to be queued when actor is disabled

another style fail of me

reverse and queueactivity activities are now executed on top of the queue for a sell action

RoosterDragon commented on OpenRA.Mods.Common/Traits/Attack/AttackTurreted.cs:
```diff
@@ -33,12 +33,15 @@ public AttackTurreted(Actor self, AttackTurretedInfo info)
-					canAttack = true;
+					return true;
 
 			return canAttack;
```
Since you return true in the loop, you can return false here and do away with the local variable.

pevers replied:
Thanks!

RoosterDragon commented:
Tested and fixes both issues. I'm not a buff on activities but these changes look good to me.

Could you squash the two optimization commits into one? :+1: from me once that is done.

pevers commented:
Thanks, squashed and rebased!

obrakmann commented:
I can confirm that this fixes #7403. I cannot reproduce #7442 at all, so can't confim the fix there. I did notice a huge regression with AttackMove and subsequent Move orders due to that fix, though.

* With AttackMove, the unit's movement isn't fluid, they stop at least once per second and then resume along their path immediately.
* After an AttackMove has finished, turreted units will ignore Move commands. They'll react to another AttackMove command, though.

pevers commented:
@obrakmann Wow, I didn't notice this. Tanks are impossible to control in this 'fix', i'll throw out this change.

RoosterDragon commented:
Perhaps you can keep the fix for #7442 if you make the check more specific: i.e. cancel the activity if it cannot attack specifically due to low power, which hopefully prevents it cancelling other important activities (just thinking aloud here, not sure if that is feasible).

pevers commented:
I may have found out the real issue and why it isn't always reproducible in https://github.com/OpenRA/OpenRA/issues/7442 (see the last comment).

fc44345 felt a little bit hacky so I fixed it by giving the reverse animation the option to execute the next activity unqueued. This is now only the case when the user is selling something and not when an MVC is deploying/packing. Maybe I miss something here, don't know but it seems to work now (at least for me).

pevers commented:
Well, pretty impressive number of commits for such a small issue. But I think this will improve the following:
    * taking over a turret will clear the current target
    * disabled buildings will not queue new attack activities
    * this will result in fixing the sell activity on a powered down building
    * I also changed WithMakeAnimation.Reverse to be able to execute the activity after the animation on top of the queue.

obrakmann commented on OpenRA.Mods.Common/Traits/Sellable.cs:
```diff
@@ -53,9 +53,9 @@ public void Sell(Actor self)
 			var makeAnimation = self.TraitOrDefault<WithMakeAnimation>();
 			if (makeAnimation != null)
-				makeAnimation.Reverse(self, new Sell());
+				makeAnimation.Reverse(self, new Sell(), true);
```
`true` is already the default value for this optional parameter, so you don't need to explicitly specify it. Or did you mean for this to be `false`?

obrakmann commented on OpenRA.Mods.Common/Traits/Sellable.cs:
```diff
@@ -53,9 +53,9 @@ public void Sell(Actor self)
+				makeAnimation.Reverse(self, new Sell(), true);
 			else
-				self.QueueActivity(new Sell());
+				self.QueueActivity(true, new Sell());
```
Same here as in the comment above, really: `QueueActivity(true, new SomeActivity())` is the same as just `QueueActivity(new SomeActivity())`. It only behaves differently if the `queued` paramter is false.

So I don't quite get what you want to achieve with this change. The behaviour with this patch is exactly the same as without it.

pevers commented:
@obrakmann thanks for te review!

That's a mistake, that must be false (I was trying if it would still work, and it did). I added the parameter animation method to be able to execute the next activity right away just as with QueueActivity. 

In this change I prevented an AttackBase activity to be scheduled if the actor is disabled. This will fix https://github.com/OpenRA/OpenRA/issues/7442 but I was still worried that maybe some other activity would jump in between the sell-animation and the actually sell. So I added the option to the QueueActivty and Render.

obrakmann commented:
Looks good now (although I still can't reproduce #7442). Please squash the commits, though.

pevers commented:
I have squashed the commits.

pevers commented:
Can the 'Changes Requested' label be removed?

obrakmann commented:
So, I've *finally* been able to reproduce that bug, and can now confirm that this is fixed with this change. :+1:

obrakmann commented:
Thanks!

[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/a6a72ace91215ff0dc9be1c65b9abe44c57cb1cc%5E...a6a72ace91215ff0dc9be1c65b9abe44c57cb1cc)
